feat: chat ffi

brianp opened the pull request:
Description
---
Create an FFI wrapper for a simple chat client


Motivation and Context
---
So we can chat on mobile and other non rust platforms.

How Has This Been Tested?
---
Cucumber

What process can a PR reviewer use to test or verify this change?
---
Review the cucumber tests as well

<!-- Checklist -->
<!-- 1. Is the title of your PR in the form that would make nice release notes? The title, excluding the conventional commit
tag, will be included exactly as is in the CHANGELOG, so please think about it carefully. -->


Breaking Changes
---

- [x] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [ ] Other - Please specify

<!-- Does this include a breaking change? If so, include this line as a footer -->
<!-- BREAKING CHANGE: Description what the user should do, e.g. delete a database, resync the chain -->

brianp committed:
Extract chat client from integration tests

If possible I'll use a common client for FFI so moving the client from
the testing suite into the contacts crate makes it more available.

brianp committed:
Commit the chat ffi base files

brianp committed:
Begin making the chat client more genericlly usable

brianp committed:
Basic client creation and msg sending via FFI

SWvheerden commented on base_layer/chat_ffi/build.rs:
```diff
@@ -0,0 +1,55 @@
+        .write_to_file(output_file);
+}
+
+// /// Find the location of the `target/` directory. Note that this may be
+// /// overridden by `cmake`, so we also need to check the `CARGO_TARGET_DIR`
+// /// variable.
+// fn target_dir() -> PathBuf {
+//     if let Ok(target) = env::var("CARGO_TARGET_DIR") {
+//         PathBuf::from(target)
+//     } else {
+//         PathBuf::from(env::var("CARGO_MANIFEST_DIR").unwrap()).join("target")
+//     }
+// }
```
delete?

SWvheerden commented on base_layer/chat_ffi/src/lib.rs:
```diff
@@ -0,0 +1,153 @@
+#[no_mangle]
+pub unsafe extern "C" fn send_message(
+    client: &mut ClientFFI,
+    receiver_c_char: *const c_char,
```
```suggestion
    receiver_c_char: *TariAdresses,
```
Rather make it take a tari address here, then we can be sure it should work here and we don't have to expect it. We can effectively copy it over from the wallet ffi. It takes tari address ppointers.

SWvheerden commented on base_layer/chat_ffi/src/lib.rs:
```diff
@@ -0,0 +1,153 @@
+    File::open(identity_path)
+        .expect("Can't open the identity file")
+        .read_to_end(&mut buf)
+        .expect("Can't read the identity file into buffer");
```
can you try and make then return some error code rather than crashing on an except?

SWvheerden reviewed (commented):
This is looking good. Only thing I think we can improve on is removal of the excepts inside of the ffi

brianp committed:
Treat all chat clients the same in cucumber

Whether FFI or native client treat them the same to reduce duplication
of similar functions.

brianp committed:
Update workflow

brianp committed:
Update remaining functions

brianp committed:
Clippy warnings

brianp committed:
Cast to generic pointers to make ffi "safer"

brianp committed:
Refactor the get_all_messages function to be less memory leaky

brianp committed:
Remove unneeded deps and clean up build file

brianp committed:
Add better error handling

SWvheerden committed:
Merge branch 'development' into chat-ffi

SWvheerden commented:
ACK
